src/main/flight/mixer.c

```diff
@@ -269,83 +273,116 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+    const float cosThreshold = sqrtf(3.0f)/2.0f; // cos(PI/6.0f)
+
+    if (cosPhi < cosThreshold) {
+        // Enforce either roll or pitch exclusively, if not on diagonal
+        if (stickDeflectionRollAbs > stickDeflectionPitchAbs) {
             signPitch = 0;
         } else {
-            // If pitch/roll dominant, disable yaw
-            signYaw = 0;
+            signRoll = 0;
         }
+    }
 
-        const float cosPhi = (stickDeflectionLength > 0) ? (stickDeflectionPitchAbs + stickDeflectionRollAbs) / (sqrtf(2.0f) * stickDeflectionLength) : 0;
-        const float cosThreshold = sqrtf(3.0f)/2.0f; // cos(PI/6.0f)
+    // Calculate crashflipPower from stick deflection with a reasonable amount of stick deadband
+    float crashflipPower = stickDeflectionLength > CRASHFLIP_STICK_MINF ? stickDeflectionLength : 0.0f;
+
+    // calculate flipPower attenuators
+    float crashflipRateAttenuator = 1.0f;
+    float crashflipAttitudeAttenuator = 1.0f;
+    const float crashflipRateLimit = mixerConfig()->crashflip_rate * 10.0f; // eg 35 = no power by 350 deg/s
+
+    // disable both attenuators if the user's crashflip_rate is zero
+    if (crashflipRateLimit > 0) {
+#ifdef USE_ACC
+        // Calculate an attenuator based on change of attitude (requires Acc)
+        // with Acc, crashflipAttitudeAttenuator will be zero after approx 90 degree rotation, and
+        // motors will stop / not spin while attitude remains more than ~90 degrees from initial attitude
+        // without Acc, the user must manually center the stick, or exit crashflip mode, or disarm, to stop the motors
+        // re-initialisation of crashFlip mode by arm/disarm is required to reset the initial tilt angle
+        if (sensors(SENSOR_ACC)) {
+            const float tiltAngle = getCosTiltAngle();  // -1 if flat inverted, 0 when 90° (on edge), +1 when flat and upright
+            if (isFirstTiltAngleRead) {
+                tiltAngleAtStart = tiltAngle;
+                isFirstTiltAngleRead = false;
+            }
+            crashflipAttitudeAttenuator = fmaxf(1.0f - fabsf(tiltAngleAtStart - tiltAngle), 0.0f);
+        }
+#endif // USE_ACC
+        // Calculate an attenuation factor based on rate of rotation... note:
+        // if driving roll or pitch, quad usually turns on that axis, but if one motor sticks, could be a diagonal rotation
+        // if driving diagonally, the turn could be either roll or pitch
+        // if driving yaw, typically one motor sticks, and the quad yaws a little then flips diagonally
+        float gyroRate = 0.0f;
+        for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+            gyroRate = fmaxf(gyroRate, fabsf(gyro.gyroADCf[axis]));
+        }
+        crashflipRateAttenuator = fmaxf((crashflipRateLimit - gyroRate) / crashflipRateLimit, 0.0f);
+        crashflipPower *= crashflipAttitudeAttenuator * crashflipRateAttenuator;
```

Comment:
I had originally done it that way. 
However, the following considerations led to the approach I now have:
- people also use crashflip to get out of trees, in situations where the quad may be upright
- Sometimes after flipping, the quad can remain a bit stuck, even though fundamentally it is upright. Pilots then sometimes stay in crashflip mode trying to get in-stuck. With the current code, a disarm/rearm cycle will re-initialise the angle and they can try that. If we only care about ‘being upright’, they cannot
- The ‘upright’ test doesn’t allow smooth attenuation of motors, starting as soon as the quad starts to reduce angle. They will get strong drive until it becomes upright. This tends to cause over-rotation.

Reply:
Okay, makes sense. I didn't think of point 1 and 2. They can be solved by using the change of `isUpright()` instead. Then the third point is just a case of lowering the `crashflip_rate` value.

And please excuse my ignorance. I guess this now boils down to this question:
Does using a **linear attitude attenuator** yield any upsides besides helping to limit the overall rotation rate (to prevent over-rotation)? I'm asking because by going **from linear to constant attitude attenuation** we can maybe make the feature usable for a more wide variety of quads.

Reply:
Hmm, I don't quite understand your comment`Okay...`.  As far as I can tell, we can't use `isUpright()` at all in the crashflip setting, since we will need full power while upright in some situations.

I haven't experimented with a linear *attitude* attenuator, and don't know how to write that sort of function.

This code can be easily tested indoors; please experiment with it and see what works best for you.

Reply:
> It's possible that, now we have gyro attenuation as well, we could just use the attitude more as an on/off switch, and use the gyro attenuation as the primary 'linearising' factor.

Yes, that's what I was after :)

It's quite simple, you already have the structure in place. Just instead of the delta of (the cosine of) the tilt angle, you'd use the delta of `isUpright()`. This means, if the quad **transitions** from "not upright to upright" or from "upright to not uprigt", the attitude attenuator would engage and stop the flip. Same basic behavior that you have in place right now, just that the attitude attenuator goes from 1 to 0 instantly, when the quad turns over on each side. This leaves the rate attenuator in full control up until the quad turned over onto its other side.

No idea if this helps but over-rotation should be no problem anymore if we dial down the `crashflip_rate`. If it's still a problem afterwards, just ignore this.